### PR TITLE
[BG-5837] Create route to provision a new wallet key

### PR DESCRIPTION
**Pull request by dacruz21**
Exposes the /key route, which derives a wallet key off of the customer/coin's master key, allocating a master key if necessary.

NOTE: I'm not yet happy with the unit tests, so I will not merge this yet. I want to test cases where the test suite is certain that a master key already exists, or does not exist, or no master keys are available. PR #8 has some nice test database construction/deconstruction code, so I'll wait for that to land and then write more tests. But I wanted to get this PR up so we can start reviewing the actual logic now.

/key expects the following parameter's in the POST request's body:
 - customerId: public ID of the enterprise or user requesting the key
 - coin: coin ticker (btc, eth, ltc...)
 - userEmail: email address of the user creating the wallet. this is not a unique identifier (that's the purpose of the customerId), but rather exists so that the KRS provider can contact the user to inform them that the key has been created, or during the recovery process
 - notificationURL: for heavy API customers, the keyserver will fire off a notification to this URL so that customers can take note of their public key without receiving an email for every wallet
 - disableKRSEmail: when set to true, will not send an email to the user's email address. usually used in tandem with notificationURL
 - custom: can be any additional information, such as phone number
 - requesterId: when running in secure mode, the platform needs to identify itself
 - requesterSecret: when running in secure mode, the platform uses this to authenticate with the keyserver

### CODE REVIEW

File: app/krs.js [hunk 1]
```diff
@@ -78,38 +85,50 @@ exports.provisionKey = function(req) {
...
-  .then(function() {
-    return notifyEndpoint(key, 'created');
-  })
-  .then(function() {
-    return key;
-  });
-};
+  }
+
+  yield notifyEndpoint(key, 'created');
```
Comment (marktoda) [inline app/krs.js]: Maybe do the check for if(notifyEndpoint == null) here, for readability and so we don't go to the trouble of generating HMAC and then just breaking without using it. Or you could turn notifyEndpoint into just notify and have both the email logic and the notifyEndpoint logic in there, switched on the flags

File: app/utils.js [hunk 2]
```diff
@@ -83,3 +85,9 @@ exports.sendMailQ = function(toEmail, subject, template, templateParams, attachm
 exports.formatBTCFromSatoshis = function(satoshis) {
   return (satoshis * 1e-8).toFixed(4);
 };
+
+exports.deriveChildKey = function(masterXpub, derivationPath) {
+  const masterNode = prova.HDNode.fromBase58(masterXpub);
```
Comment (marktoda) [inline app/utils.js]: I don't think this is coin agnostic. Also, use a different library like bitgo-utxo-lib and equivalent

File: app/utils.js [hunk 3]
```diff
@@ -83,3 +85,9 @@ exports.sendMailQ = function(toEmail, subject, template, templateParams, attachm
...
+
+  return masterNode.derivePath(derivationPath).toBase58();
```
Comment (marktoda) [inline app/utils.js]: Same, make coin agnostic

File: app/utils.js (hunk 2 above)
Comment (dacruz21) [reply app/utils.js]: As discussed over WhatsApp, this is fine for all secp256k1 coins. bitgo-utxo-lib shouldn't be necessary. Also, we use xpubs for all networks, no need to worry about tpubs, etc.

File: app/utils.js (hunk 3 above)
Comment (dacruz21) [reply app/utils.js]: Same as above

File: app/krs.js [hunk 4]
```diff
@@ -78,38 +110,52 @@ exports.provisionKey = function(req) {
...
+  }
+
+  key.masterKey = masterKey.xpub;
+  key.path = `m/${masterKey.keyCount}`;
+  key.xpub = utils.deriveChildKey(key.masterKey, key.path);
+
+  key.userEmail = req.body.userEmail;
+  key.notificationURL = req.body.notificationURL;
+
+  key.custom = req.body.custom || {};
```
Comment (marktoda) [inline app/krs.js]: I'm confused -- in www, the old krs code called it "userCustom" -- I'd like to keep the same naming convention to minimize the amount of changed code for sending basically the same data. Confusion comes from this file looking like it used to call it just "custom" too -- maybe www is wrong and it's just always come through as null lol

Comment (marktoda) [reply app/krs.js]: worked this out with david off band

**Pull request merged**